File: android/app/src/main/java/com/reactnativebridge/NativeModule/NativeModule.kt
package com.reactnativebridge

import android.util.Log
import com.facebook.react.bridge.Promise
import com.facebook.react.bridge.ReactApplicationContext
import com.facebook.react.bridge.ReactContextBaseJavaModule
import com.facebook.react.bridge.ReactMethod


class NativeModule(context: ReactApplicationContext) : ReactContextBaseJavaModule(context) {

    override fun getName(): String {
        return "NativeModule"
    }

    @ReactMethod
    fun logMessage(message: String) {
        Log.d("MyApp", message)
    }

    @ReactMethod
    fun add(a: Int, b: Int, promise: Promise) {
        val result = a + b
        promise.resolve(result)
    }

    @ReactMethod
    fun subtract(a: Int, b: Int, promise: Promise) {
        if (a >= b) {
            promise.resolve(a - b)
        } else {
            val error = Exception("a is less than b")
            promise.reject("subtract_error", "a is less than b", error)
        }
    }
}

File: ios/NativeModule/NativeModule.swift
//
//    .swift
//  ReactNativeBridge
//
//  Created by Nayan Vekariya on 18/02/25.
//

import Foundation
import React

@objc(NativeModule)
class NativeModule: NSObject {

  @objc static func requiresMainQueueSetup() -> Bool {
      return true
  }

  @objc func logMessage(_ message: String) {
    NSLog("%@", message)
  }

  @objc func add(_ a: Int, b: Int, resolve: @escaping RCTPromiseResolveBlock, reject: @escaping RCTPromiseRejectBlock) {
      if a >= 0 && b >= 0 {
          resolve(a + b)
      } else {
          let error = NSError(domain: "AddError", code: 400, userInfo: [NSLocalizedDescriptionKey: "Both numbers must be non-negative"])
          reject("add_error", "Invalid input: a and b must be non-negative", error)
      }
  }

  @objc func subtract(_ a: Int, b: Int, resolve: RCTPromiseResolveBlock, reject: RCTPromiseRejectBlock) {
    if a >= b {
      resolve(a - b)
    } else {
      let error = NSError(domain: "SubtractError", code: 200, userInfo: nil)
      reject("subtract_error", "a is less than b", error)
    }
  }
}

